### PR TITLE
Implement flag to allow typechecking of untyped modules

### LINKED ISSUE #8545: Allow/disallow pep561 packages
@alexander255 has requested a flag `--pep561-override` to allow listing packages as pep 561 compatible when they are not. This got me thinking and I think the best UX for this would be a pair of flags: `--allow-installed-package NAME` and `--disallow-installed-package NAME` (or something like that).

The idea is to solve two problems:

Some people publish PEP 561 packages that are either too unstable or flat out broken for some users. This allows the user to disable the package until things are fixed.

At the other end of the spectrum, some packagers are conservative on adding types, and while they exist, they are not public (yet). Some users would still like to use them.

This would eliminate `--no-site-packages`, which is a rather blunt instrument.

### PR DESCRIPTION
**Pull request by DeinAlptraum**
Add a flag and config ini options `"follow_untyped_imports"`. Setting it to `True` instructs mypy to typecheck also modules that do not have stubs or a `py.typed` marker.

Fixes #8545

### CODE REVIEW

**Comment by DeinAlptraum** [comment]
@hauntsaninja can you tell me by any chance if there's anything I'm missing/I should do to get a review here?

**Comment by DeinAlptraum** [comment]
Ping. @hauntsaninja @JelleZijlstra can you tell me what I need to do to get a review here?

**Comment by DeinAlptraum** [comment]
Ping @hauntsaninja @JelleZijlstra @JukkaL

**Comment by JelleZijlstra** [comment]
Mypy development is unfortunately under-resourced and personally I don't have much time available to review PRs. I'd advise you to wait and hopefully a committer will come along and review this.

**Comment by Xiddoc** [comment]
I feel like this feature is crucial and affects way too many people for it to be ignored. Using `type: ignore` or manually adding a `py.typed` file to every used third-party module is incredibly impractical.

Could this at least get a CR, if improvements are needed then the community can follow up on issues for the PR. ❤️ 
@JukkaL @hauntsaninja

**Comment by JukkaL** [comment]
I can try to review this -- possibly next week. It would help if the merge conflict was fixed.

**Comment by DeinAlptraum** [comment]
Thanks for the comment, didn't notice the merge conflict. I wish there were notifications for this... anyway, fixed.

**Review by hauntsaninja — COMMENTED** [review]
Thanks! I think the per-module handling isn't quite right, in particular, I'm not sure it handles globs correctly. I'd also vote for a slightly more specific name, maybe something like `--follow-untyped-imports`. This matches the untyped-import error code it will silence.

**Comment by DeinAlptraum** [comment]
@hauntsaninja thanks for the feedback! You're right, this wasn't handling globs properly. In the process of fixing that, I also noticed that my solution was more complicated than necessary and simplified it  a bit...

Regarding the name: it seems there was quite a lot of discussion about how the config option should be called on the original issue, so I wasn't sure what to pick. Yours definitely sounds better though, so changed it to that for now.

File: mypy/modulefinder.py
Comment (Xiddoc) [inline mypy/modulefinder.py:357]: No need for this comment

Comment (DeinAlptraum) [reply mypy/modulefinder.py]: Thanks for pointing this out, overlooked it when I removed my debugging statements.

**Review by hauntsaninja — APPROVED** [review]
Thanks, this looks good. I'll merge in a few days if no one else has comments. Are you up for adding docs for this flag as well?

**Comment by DeinAlptraum** [comment]
@hauntsaninja I added it to the docs where it felt sensible to me. Do you think it should be mentioned in any other places?

**Review by hauntsaninja — APPROVED** [review]
Looks excellent, thank you for this feature!

**Comment by cdce8p** [comment]
I tried this flag for Home Assistant today. Unfortunately, mypy couldn't even do a full run and failed instantly. Haven't narrowed it down completely yet, just a first reproducer (tested with `Python 3.13`):
```py
# test.py
import aprslib.base91
reveal_type(1)
```
```ini
# mypy_config.ini
[mypy]
follow_imports = normal
follow_untyped_imports = true
```
```bash
pip install aprslib==0.7.2
mypy --config-file mypy_config.ini test.py
```
```py
/.../venv/lib/python3.13/site-packages/aprslib/parsing/common.py:218: SyntaxWarning: invalid escape sequence '\!'
  match = re.findall("^(.*)\!([\x21-\x7b])([\x20-\x7b]{2})\!(.*?)$", body)
venv/lib/python3.13/site-packages/aprslib/parsing/__init__.py: error:
    Source file found twice under different module names: "aprslib.parsing" and "aprslib.parsing.__init__"
Found 1 error in 1 file (errors prevented further checking)
```

_The `SyntaxWarning` is an unrelated issue._

Will open a new issue, once I've narrowed it down a bit more.

**Comment by DeinAlptraum** [comment]
@cdce8p thanks for reporting, but that seems unrelated to this change.
I just cloned the aprs-python repo and ran the latest mypy release (i.e. before I added this flag) in the root folder of the project and I get the same error. Couldn't really tell you what that means though, since I'm not super familiar with the inner workings of mypy

**Comment by cdce8p** [comment]
> @cdce8p thanks for reporting, but that seems unrelated to this change. I just cloned the aprs-python repo and ran the latest mypy release (i.e. before I added this flag) in the root folder of the project and I get the same error.

Indeed thanks for pointing it out! Thinking out loud for a moment: Would it make sense to add an explicit warning to the docs or am I just the only one who would run into it? With `follow_untyped_imports = true` it's much more likely that code will be checked which isn't designed for type checking so errors might occur more frequently.

**Comment by hauntsaninja** [comment]
Possibly same thing as https://github.com/python/mypy/issues/9716 ?

**Comment by hauntsaninja** [comment]
Yeah looks like that might be right: https://github.com/rossengeorgiev/aprs-python/blob/master/aprslib/parsing/thirdparty.py#L2

**Comment by hauntsaninja** [comment]
We can add a warning to the docs, done so here: https://github.com/python/mypy/pull/18249

**Pull request merged**
